### PR TITLE
feat(049-human-in-loop): Complete human-in-the-loop blocker system (Phases 1-9)

### PR DESCRIPTION
**Pull request by frankbria**
## Summary

Implements complete human-in-the-loop blocker system enabling CodeFRAME agents to request human assistance when blocked, pause work, receive answers via dashboard, and automatically resume execution with incorporated context.

### Feature Overview
- **5 User Stories Delivered**: Agent blocker creation, dashboard resolution UI, agent resume workflow, SYNC/ASYNC handling, webhook notifications
- **ALL 70 Tasks Complete**: Phases 1-10 fully implemented (Setup, Foundational, US1-US5, Blocker Expiration, Testing, Polish)
- **42 Tests Passing**: 100% pass rate on unit, integration, WebSocket, and quickstart validation tests

---

### ✅ US1: Agent Blocker Creation and Display (P1)
- Agents can create blockers via `create_blocker()` method (BackendWorkerAgent, FrontendWorkerAgent, TestWorkerAgent)
- Real-time blocker display in dashboard BlockerPanel with WebSocket updates
- GET /api/projects/:project_id/blockers and GET /api/blockers/:blocker_id endpoints

### ✅ US2: Blocker Resolution via Dashboard (P1)
- BlockerModal component for viewing full blocker details and submitting answers
- POST /api/blockers/:blocker_id/resolve endpoint with conflict handling (409)
- Answer validation (non-empty, max 5000 chars with character counter)
- Success/error toast notifications for user feedback

### ✅ US3: Agent Resume After Resolution (P1)
- `wait_for_blocker_resolution()` polling method in all worker agents
- Automatic answer injection into agent task context
- agent_resumed WebSocket event with dashboard status updates

### ✅ US4: SYNC vs ASYNC Blocker Handling (P2)
- SYNC blockers: Pause dependent tasks (red/CRITICAL badge)
- ASYNC blockers: Allow parallel work (yellow/INFO badge)
- LeadAgent dependency handling for task coordination

### ✅ US5: Blocker Notifications (P3)
- Webhook notification service for SYNC blockers
- BLOCKER_WEBHOOK_URL environment variable configuration
- Async fire-and-forget delivery with 5s timeout and error logging

---

### Database
- Migration 003: Updated blockers table schema (blocker_type, status, timestamps)
- 7 database methods: create_blocker(), resolve_blocker(), get_pending_blocker(), list_blockers(), get_blocker(), expire_stale_blockers(), get_blocker_metrics()
- Duplicate resolution prevention with status-based locking
- Rate limiting: 10 blockers/minute per agent

### Backend API (FastAPI)
- 4 RESTful endpoints with comprehensive error handling (404, 409, 422)
- WebSocket broadcasts: blocker_created, blocker_resolved, agent_resumed, blocker_expired
- Webhook notification integration for external alerting
- Blocker metrics endpoint for analytics

### Frontend (React + TypeScript)
- 3 components: BlockerPanel, BlockerModal, BlockerBadge
- Real-time WebSocket event handling
- Filtering: All / SYNC / ASYNC with toggle buttons
- Sorting: SYNC first, then by created_at DESC (newest first)
- API client methods in web-ui/src/lib/api.ts
- Tailwind CSS styling with SYNC/ASYNC visual distinction

### Testing
- **Unit Tests (20)**: Blocker CRUD operations, expiration, polling, concurrent resolution
- **Integration Tests (22)**: End-to-end workflows, SYNC/ASYNC handling, edge cases, quickstart validation
- **Frontend Tests**: WebSocket event integration, component rendering
- **Pass Rate**: 42/42 tests passing (100%)

---

## Stale Blocker Expiration

- Automatic 24-hour expiration for unresolved blockers
- Hourly cron job: codeframe/tasks/expire_blockers.py
- Task failure logic when blocker expires (update task status to FAILED)
- blocker_expired WebSocket event for dashboard updates

---

## Phase 10: Polish & Cross-Cutting Concerns

**ALL Tasks Complete** (T062-T070):
- ✅ T062: Blocker metrics tracking (avg resolution time, expiration rate, counts by status/type)
- ✅ T063: Rate limiting (10 blockers/minute per agent to prevent spam)
- ✅ T064: Comprehensive error handling (404, 409, 422 status codes)
- ✅ T065: Character counter in BlockerModal (4500/5000)
- ✅ T066: Conflict handling (409 → "Already resolved by another user")
- ✅ T067: BlockerPanel sorting (SYNC first, then by created_at DESC)
- ✅ T068: BlockerPanel filtering (All / SYNC / ASYNC toggle buttons)
- ✅ T069: Quickstart validation (12 tests covering tutorial, patterns, troubleshooting)
- ✅ T070: Complete documentation and docstrings

---

### Backend Tests
```
tests/test_blockers.py:                        20 passed  (0.85s)
tests/integration/test_blocker_workflow.py:    10 passed  (0.45s)
tests/integration/test_quickstart_validation.py: 12 passed  (0.51s)
Total: 42/42 tests passing (100%)
```

### Frontend Tests
```
web-ui/__tests__/integration/blocker-websocket.test.ts: All WebSocket events validated
web-ui/__tests__/components/BlockerPanel.test.tsx: Component rendering verified
web-ui/__tests__/components/BlockerBadge.test.tsx: Badge display validated
```

---

## Commit History

- Phase 1-2: Database schema and foundational infrastructure (T001-T010)
- Phase 3: User Story 1 - Agent blocker creation and dashboard display (T011-T020)
- Phase 4: User Story 2 - Dashboard blocker resolution (T021-T027)
- Phase 5: User Story 3 - Agent resume after resolution (T028-T034)
- Phase 6: User Story 4 - SYNC/ASYNC blocker handling (T035-T039)
- Phase 7: User Story 5 - Webhook notifications (T040-T044)
- Phase 8: Stale blocker expiration (T045-T049)
- Phase 9: Comprehensive testing (T050-T061)
- Phase 10: Polish, metrics, rate limiting, filtering, validation (T062-T070)

---

## Documentation

- **Specification**: specs/049-human-in-loop/spec.md
- **Architecture**: specs/049-human-in-loop/plan.md
- **Data Model**: specs/049-human-in-loop/data-model.md
- **Tasks**: specs/049-human-in-loop/tasks.md
- **Quickstart**: specs/049-human-in-loop/quickstart.md

---

## Breaking Changes

None - Feature is additive and does not modify existing functionality.

---

## Migration Required

Yes - Run migration 003 before deploying:
```bash
# Migration will be applied automatically on server startup
# Or manually: python -m codeframe.persistence.migrations.migration_003_update_blockers_schema
```

---

## Environment Variables

Optional webhook configuration:
```bash
BLOCKER_WEBHOOK_URL=https://your-webhook-endpoint.com/blockers
```

---

## Reviewers

@frankbria

---

## Checklist

- [x] All user stories (US1-US5) implemented and tested
- [x] Database migration created and tested
- [x] API endpoints tested with error handling
- [x] Frontend components tested with WebSocket integration
- [x] 42/42 tests passing (100% pass rate)
- [x] Documentation complete (docstrings, quickstart, specs)
- [x] Phase 10 polish complete (metrics, rate limiting, filtering, validation)
- [x] Ready for production deployment

## Summary by CodeRabbit

* **New Features**
  * Human-in-the-loop blockers: SYNC pauses dependents, ASYNC is informational; create/list/view/resolve via API
  * UI: Blocker panel, modal for resolving blockers, dashboard integration with real-time updates and badges
  * Blocker metrics and webhook notifications for critical (SYNC) blockers

* **Tools**
  * Cron job to expire stale blockers and mark affected tasks failed; rate limiting on blocker creation

* **Tests**
  * Extensive unit and integration tests covering lifecycle, expiration, webhooks, and UI components

### CODE REVIEW

File: codeframe/agents/frontend_worker_agent.py [hunk 1]
```diff
@@ -414,4 +415,272 @@ def _update_imports_exports(
...
+
+        # 3. Inject answer into context
+        enriched_context = {
+            **context,
+            "blocker_answer": answer,
+            "blocker_question": question,
+            "blocker_id": blocker_id
+        }
+
+        return enriched_context
```
[the remark this answers is not in the document]
Comment (frankbria) [reply codeframe/agents/frontend_worker_agent.py]: @claude Does this need addressing?

File: codeframe/tasks/expire_blockers.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                    )
+                    logger.debug(f"Broadcast blocker_expired event for blocker {blocker_id}")
+                except Exception as e:
+                    logger.warning(f"Failed to broadcast blocker_expired event: {e}")
+
+        return len(expired_ids)
+
+    finally:
+        db.close()
+
```
[the remark this answers is not in the document]
Comment (frankbria) [reply codeframe/tasks/expire_blockers.py]: @claude Does this need addressing?

File: codeframe/agents/frontend_worker_agent.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (frankbria) [reply codeframe/agents/frontend_worker_agent.py]: @claude Yes, let's use the better approach.

File: codeframe/agents/test_worker_agent.py [hunk 3]
```diff
@@ -595,4 +596,273 @@ async def _broadcast_test_result(
...
+                    )
+                    logger.debug(f"Webhook notification queued for SYNC blocker {blocker_id}")
+                else:
+                    logger.debug("BLOCKER_WEBHOOK_URL not configured, skipping webhook notification")
+
+            except Exception as e:
+                # Log error but don't block blocker creation
+                logger.warning(f"Failed to send webhook notification for blocker {blocker_id}: {e}")
+
+        return blocker_id
```
[the remark this answers is not in the document]
Comment (frankbria) [reply codeframe/agents/test_worker_agent.py]: @claude yes, let's make those fixes as well.

File: tests/test_webhook_notifications.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+                question="Critical blocker",
+                agent_id="backend-worker-1",
+                task_id=456,
+                blocker_type=BlockerType.SYNC,
+                created_at=created_at
+            )
+
+        assert result is True
+        mock_session.post.assert_called_once()
+
```
[the remark this answers is not in the document]
Comment (frankbria) [reply tests/test_webhook_notifications.py]: @claude Does this need a fix?
